prov/gni/src/gnix_rma.c

```diff
@@ -0,0 +1,245 @@
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include "gnix.h"
+#include "gnix_nic.h"
+#include "gnix_vc.h"
+#include "gnix_ep.h"
+#include "gnix_mr.h"
+#include "gnix_cm_nic.h"
+#include "gnix_mbox_allocator.h"
+#include <gni_pub.h>
+
+static int __gnix_rma_fab_req_complete(void *arg)
+{
+	struct gnix_fab_req *req = (struct gnix_fab_req *)arg;
+	struct gnix_fid_ep *ep = req->gnix_ep;
+	int rc;
+
+	/* more transaction needed for request? */
+
+	/* write completions */
+	if (ep->send_cq && (!ep->send_selective_completion ||
+			    (req->flags & FI_COMPLETION))) {
+		rc = _gnix_cq_add_event(ep->send_cq, req->user_context,
+					req->flags, req->len,
+					(void *)req->loc_addr,
+					req->data, req->tag);
+		if (rc) {
+			GNIX_WARN(FI_LOG_CQ,
+				  "_gnix_cq_add_event() failed: %d\n", rc);
+		}
+	}
+
+	_fr_free(ep, req);
+
+	return FI_SUCCESS;
+}
+
+static int __gnix_rma_txd_complete(void *arg)
+{
+	struct gnix_tx_descriptor *txd = (struct gnix_tx_descriptor *)arg;
+
+	/* Progress fabric operation in the fab_req completer.  Can we call
+	 * fab_req->completer_func directly from __gnix_tx_progress? */
+	return txd->desc.req->completer_func(txd->desc.req->completer_data);
+}
+
+static gni_post_type_t __gnix_fr_post_type(int fr_type)
+{
+	switch (fr_type) {
+	case GNIX_FAB_RQ_RDMA_WRITE:
+	case GNIX_FAB_RQ_RDMA_WRITE_IMM_DATA:
+		return GNI_POST_RDMA_PUT;
+	case GNIX_FAB_RQ_RDMA_READ:
+		return GNI_POST_RDMA_GET;
+	default:
+		break;
+	}
+
+	GNIX_WARN(FI_LOG_EP_DATA, "Unsupported post type: %d", fr_type);
+	return -FI_ENOSYS;
+}
+
+int _gnix_post_req(struct gnix_fab_req *fab_req)
```

Comment:
We may need to expose it later when sending queued requests.  I'll hide it for now